### PR TITLE
Add support for IP/L4 checksum verification.

### PR DESCRIPTION
**Pull request by ajamshed**
L4Checksum and IPCheckum verification support is enabled once the
argument 'verify=True' is passed during module initialization.
Default value is set to False. Gate 0 gets all packets that pass
the verification test. Gate 1 gets all the packets that have
incorrect checksums.

Signed-off-by: Muhammad Asim Jamshed <muhammad.jamshed@intel.com>
Signed-off-by: Saikrishna Edupuganti <saikrishna.edupuganti@intel.com>

### CODE REVIEW

File: core/modules/l4_checksum.cc [hunk 1]
```diff
@@ -49,28 +51,42 @@ void L4Checksum::ProcessBatch(Context *ctx, bess::PacketBatch *batch) {
...
-    continue;
   }
-
-  RunNextModule(ctx, batch);
 }
-
+/*----------------------------------------------------------------------------------*/
+CommandResponse L4Checksum::Init(const bess::pb::L4ChecksumArg &arg) {
+  if (arg.verify())
+    verify = arg.verify();
```
Comment (melvinw) [inline core/modules/l4_checksum.cc:87]: Assignment is fine here. Protobuf defaults are always the "zero" value when using these getters
```suggestion
verify = arg.verify();
```

File: core/modules/l4_checksum.cc [hunk 2]
```diff
@@ -49,28 +51,42 @@ void L4Checksum::ProcessBatch(Context *ctx, bess::PacketBatch *batch) {
...
+	 tcp->checksum = CalculateIpv4TcpChecksum(*ip, *tcp);
     }
-
-    continue;
   }
-
-  RunNextModule(ctx, batch);
 }
-
+/*----------------------------------------------------------------------------------*/
```
Comment (melvinw) [inline core/modules/l4_checksum.cc:84]: same style nit as in ip_checksum.cc

File: core/modules/l4_checksum.cc [hunk 3]
```diff
@@ -49,28 +51,42 @@ void L4Checksum::ProcessBatch(Context *ctx, bess::PacketBatch *batch) {
...
+CommandResponse L4Checksum::Init(const bess::pb::L4ChecksumArg &arg) {
+  if (arg.verify())
+    verify = arg.verify();
+  return CommandSuccess();
+}
+/*----------------------------------------------------------------------------------*/
```
Comment (melvinw) [inline core/modules/l4_checksum.cc:90]: same style nit as in ip_checksum.cc

File: core/modules/ip_checksum.cc [hunk 4]
```diff
@@ -34,6 +34,8 @@
 #include "../utils/ether.h"
 #include "../utils/ip.h"
 
+enum { DEFAULT_GATE = 0, _DROP_GATE };
```
Comment (melvinw) [inline core/modules/ip_checksum.cc:37]: `_DROP_GATE` could get easily confused with `DROP_GATE` used by the scheduler to discard packets. Perhaps rename these to `VALID_GATE` (or `DEFAULT_GATE`) and `INVALID_GATE`?

File: core/modules/ip_checksum.h [hunk 5]
```diff
@@ -39,7 +39,14 @@ class IPChecksum final : public Module {
  public:
   IPChecksum() : Module() { max_allowed_workers_ = Worker::kMaxWorkers; }
 
+  /* Gates: (0) Default, (1) Forward */
```
Comment (melvinw) [inline core/modules/ip_checksum.h:42]: Could you move/copy this to the argument protobuf comment?

File: core/modules/l4_checksum.cc [hunk 6]
```diff
@@ -36,6 +36,8 @@
 #include "../utils/tcp.h"
 #include "../utils/udp.h"
 
+enum { DEFAULT_GATE = 0, _DROP_GATE };
```
Comment (melvinw) [inline core/modules/l4_checksum.cc:39]: `_DROP_GATE` could get easily confused with `DROP_GATE` used by the scheduler to discard packets. Perhaps rename these to `VALID_GATE` (or `DEFAULT_GATE`) and `INVALID_GATE`?

File: core/modules/ip_checksum.cc (hunk 4 above)
Comment (ajamshed) [reply core/modules/ip_checksum.cc]: `INVALID_GATE` is also already defined. Renamed the gates.

File: core/modules/l4_checksum.cc [hunk 7]
```diff
@@ -49,27 +51,40 @@ void L4Checksum::ProcessBatch(Context *ctx, bess::PacketBatch *batch) {
     Ethernet *eth = batch->pkts()[i]->head_data<Ethernet *>();
 
     // Calculate checksum only for IPv4 packets
-    if (eth->ether_type != be16_t(Ethernet::Type::kIpv4))
+    if (eth->ether_type != be16_t(Ethernet::Type::kIpv4)) {
+      EmitPacket(ctx, batch->pkts()[i], FORWARD_GATE);
```
Comment (melvinw) [inline core/modules/l4_checksum.cc:55]: To be safe, could you include a note in the protobuf about how non-IP packets are simply forwarded with no modification.

File: core/modules/ip_checksum.cc [hunk 8]
```diff
@@ -54,7 +56,8 @@ void IPChecksum::ProcessBatch(Context *ctx, bess::PacketBatch *batch) {
       data = qinq + 1;
       ether_type = qinq->ether_type;
       if (ether_type != be16_t(Ethernet::Type::kVlan)) {
-        continue;
+	EmitPacket(ctx, batch->pkts()[i], FORWARD_GATE);
+	continue;
```
Comment (melvinw) [inline core/modules/ip_checksum.cc:60]: To be safe, could you include a note in the protobuf about how non-IP packets are simply forwarded with no modification.

also, nit: indentation got messed up here

**Comment by melvinw** [comment]
Oh and @ajamshed, add yourself to the contributors list [here](https://github.com/NetSys/bess/blob/master/CONTRIBUTING.md#list-of-contributors) if you like :)

**Comment by melvinw** [comment]
@ajamshed can you pull the latest master and rebase your branch onto it? Github can't automatically resolve the conflict on the contributors list for some reason...

**Comment by melvinw** [comment]
Thanks!

cat

**Pull request merged**